File: 01-Programming-Languages/PHP/Data-Structures/Tree/binary_tree.php
<?php
class TreeNode {
    public $value;
    public $left;
    public $right;

    public function __construct($value) {
        $this->value = $value;
        $this->left = null;
        $this->right = null;
    }
}

class BinaryTree {
    private $root;

    public function __construct() {
        $this->root = null;
    }

    public function insert($value) {
        $newNode = new TreeNode($value);

        if ($this->root === null) {
            $this->root = $newNode;
            return;
        }

        $current = $this->root;
        while (true) {
            if ($value < $current->value) {
                if ($current->left === null) {
                    $current->left = $newNode;
                    return;
                }
                $current = $current->left;
            } elseif ($value > $current->value) {
                if ($current->right === null) {
                    $current->right = $newNode;
                    return;
                }
                $current = $current->right;
            } else {
                return; // Duplicate
            }
        }
    }

    public function inorder($node = null) {
        if ($node === null) {
            $node = $this->root;
        }
        if ($node !== null) {
            $this->inorder($node->left);
            echo $node->value . " ";
            $this->inorder($node->right);
        }
    }

    public function search($value) {
        $current = $this->root;
        while ($current !== null) {
            if ($value === $current->value) {
                return true;
            }
            $current = $value < $current->value ? $current->left : $current->right;
        }
        return false;
    }
}

$tree = new BinaryTree();
foreach ([50, 30, 70, 20, 40, 60, 80] as $v) {
    $tree->insert($v);
}

echo "Inorder: ";
$tree->inorder();
echo "\n";

echo "Search 40: " . ($tree->search(40) ? "true" : "false") . "\n";
echo "Search 90: " . ($tree->search(90) ? "true" : "false") . "\n";
?>

File: 01-Programming-Languages/Swift/Data-Structures/Tree/binary_tree.swift
class TreeNode {
    var value: Int
    var left: TreeNode?
    var right: TreeNode?
    
    init(value: Int) {
        self.value = value
        self.left = nil
        self.right = nil
    }
}

class BinaryTree {
    var root: TreeNode?
    
    func insert(_ value: Int) {
        let newNode = TreeNode(value: value)
        
        guard let root = root else {
            self.root = newNode
            return
        }
        
        var current = root
        while true {
            if value < current.value {
                if let left = current.left {
                    current = left
                } else {
                    current.left = newNode
                    return
                }
            } else if value > current.value {
                if let right = current.right {
                    current = right
                } else {
                    current.right = newNode
                    return
                }
            } else {
                return // Duplicate
            }
        }
    }
    
    func inorder(_ node: TreeNode? = nil) {
        let currentNode = node ?? root
        guard let node = currentNode else { return }
        
        inorder(node.left)
        print("\(node.value) ", terminator: "")
        inorder(node.right)
    }
    
    func search(_ value: Int) -> Bool {
        var current = root
        while let node = current {
            if value == node.value { return true }
            current = value < node.value ? node.left : node.right
        }
        return false
    }
}

let tree = BinaryTree()
[50, 30, 70, 20, 40, 60, 80].forEach { tree.insert($0) }

print("Inorder: ", terminator: "")
tree.inorder()
print()

print("Search 40:", tree.search(40))
print("Search 90:", tree.search(90))